noir_stdlib/src/ec/montcurve.nr

```diff
@@ -0,0 +1,380 @@
+        fn msm<N>(self, n: [Field; N], p: [Point; N]) -> Point {
+            let mut out = Point::zero();
+
+            for i in 0..n.len() {
+                out = self.add(out, self.mul(n[i], p[i]));
+            }
+
+            out
+        }
+
+        // Point subtraction
+        fn subtract(self, p1: Point, p2: Point) -> Point {
+            self.add(p1, p2.negate())
+        }
+
+        // Conversion to equivalent Twisted Edwards curve
+        fn into_tecurve(self) -> TECurve {
+            let Self {j, k, gen} = self;
+            TECurve::new((j+2)/k, (j-2)/k, gen.into_tecurve())
+        }
+
+        // Conversion to equivalent Short Weierstraß curve
+        fn into_swcurve(self) -> SWCurve {
+            let j = self.j;
+            let k = self.k;
+            let a0 = (3-j*j)/(3*k*k);
+            let b0 = (2*j*j*j - 9*j)/(27*k*k*k);
+
+            SWCurve::new(a0, b0, self.map_into_swcurve(self.gen))
+        }
+
+        // Point mapping into equivalent Short Weierstraß curve
+        fn map_into_swcurve(self, p: Point) -> SWPoint {
+            if p.is_zero() == true {
+                SWPoint::zero()
+            } else {
+                SWPoint::new((3*p.x + self.j)/(3*self.k),
+                             p.y/self.k)
+            }
+        }
+
+        // Point mapping from equivalent Short Weierstraß curve
+        fn map_from_swcurve(self, p: SWPoint) -> Point {
+            let SWPoint {x, y, infty} = p;
+            let j = self.j;
+            let k = self.k;
+            
+            Point {x: (3*k*x - j)/3, y: y*k, infty}
+        }
+
+        // Elligator 2 map-to-curve method; see <https://datatracker.ietf.org/doc/id/draft-irtf-cfrg-hash-to-curve-06.html#name-elligator-2-method>.
+        fn elligator2_map(self, u: Field) -> Point {
+            let j = self.j;
+            let k = self.k;
+            let z = ZETA; // Non-square Field element required for map
+            
+            // Check whether curve is admissible
+            constrain j != 0;
+            let l = (j*j - 4)/(k*k);
+            constrain l != 0;
+            constrain is_square(l) == false;
+
+            let x1 = safe_inverse(1+z*u*u)*(0 - (j/k));
+            
+            let gx1 = x1*x1*x1 + (j/k)*x1*x1 + x1/(k*k);
+            let x2 = 0 - x1 - (j/k);
+            let gx2 = x2*x2*x2 + (j/k)*x2*x2 + x2/(k*k);
+
+            let x = if is_square(gx1) == true { x1 } else { x2 };
+
+            let y = if is_square(gx1) == true {
+                let y0 = sqrt(gx1);
+                if y0.sgn0() == 1 { y0 } else { 0 - y0 }
+            } else {
+                let y0 = sqrt(gx2);
+                if y0.sgn0() == 0 { y0 } else { 0 - y0 }
+            };
+
+            Point::new(x*k, y*k)
+                
+        }
+
+        // SWU map-to-curve method (via rational map)
+        fn swu_map(self, z: Field, u: Field) -> Point {
```

Comment:
Shouldn't z be fixed to a constant satisfying the conditions here https://datatracker.ietf.org/doc/id/draft-irtf-cfrg-hash-to-curve-06.html#section-6.6.2
rather than a parameter?

Reply:
It should indeed. At the time of this PR, it wasn't possible to compute the constant in Noir without adding unnecessary constraints, which is why I left it as a parameter here, unlike the case of the Elligator 2 method whose constant depends only on the field and not on the curve. Now that unconstrained functions are a thing, this should be easy to resolve.